pkg/tfbridge/token.go

```diff
@@ -121,40 +121,55 @@ func tokensKnownModules[T ResourceInfo | DataSourceInfo](
 // subject to change without warning.
 func TokensKnownModules(
 	tfPackagePrefix, defaultModule string, modules []string, finalize MakeToken,
-) (ResourceStrategy, DatasourceStrategy) {
+) ComputeDefaultInfo {
 	// NOTE: We could turn this from a sort + linear lookup into a radix tree to recover
 	// O(log(n)) performance (current is O(n*m)) where n = number of modules and m =
 	// number of mappings.
 	sort.Sort(sort.Reverse(sort.StringSlice(modules)))
 
-	return tokensKnownModules(tfPackagePrefix, defaultModule, modules, func(mod, tk string) (*ResourceInfo, error) {
-			tk, err := finalize(mod, tk)
-			if err != nil {
-				return nil, err
-			}
-			return &ResourceInfo{Tok: tokens.Type(tk)}, nil
-		}), tokensKnownModules(tfPackagePrefix, defaultModule, modules, func(mod, tk string) (*DataSourceInfo, error) {
-			tk, err := finalize(mod, "get"+tk)
-			if err != nil {
-				return nil, err
-			}
-			return &DataSourceInfo{Tok: tokens.ModuleMember(tk)}, nil
-		})
+	return ComputeDefaultInfo{
+		Resource: tokensKnownModules(tfPackagePrefix, defaultModule, modules,
+			func(mod, tk string) (*ResourceInfo, error) {
+				tk, err := finalize(mod, tk)
+				if err != nil {
+					return nil, err
+				}
+				return &ResourceInfo{Tok: tokens.Type(tk)}, nil
+			}),
+		DataSource: tokensKnownModules(tfPackagePrefix, defaultModule, modules,
+			func(mod, tk string) (*DataSourceInfo, error) {
+				tk, err := finalize(mod, "get"+tk)
+				if err != nil {
+					return nil, err
+				}
+				return &DataSourceInfo{Tok: tokens.ModuleMember(tk)}, nil
+			}),
+	}
+}
+
+func (ts ComputeDefaultInfo) Unmappable(substring, reason string) ComputeDefaultInfo {
```

Comment:
In practice, marking something as unmappable without explaining why leads to confusion later down the line. I'm augmenting this API to allow the user to explain why something is unmappable.

Reply:
SGTM.